### ISSUE TITLE
Sample payloadImage -- Will we provide it for every release?

### ISSUE THREAD

**Issue by fidencio** [confidential-containers/community#63]
It seems that providing a "sample payloadImage" was something added to the scope of the release as a last minute thing, therefore there was absolutely zero tooling for doing this in an automated way, leading to an image that's not easily reproducible to build.

We need to have an agreement on whether this image should be part of **every single release**, and if so, create the appropriate issues (and fixes) on the projects providing this payload to have it generated in an automated manner.

**Comment by stevenhorsman** [comment]
So to flesh this out a little. The approach that I took for 0.1 (on the day before release), was to just create a [custom branch on my fork](https://github.com/kata-containers/kata-containers/compare/CCv0...stevenhorsman:kata-containers:sample-ssh-demo-payload?expand=1) with the ssh-demo encrypted image config hard-coded into the rootfs.sh and then I ran `make cc-payload` on that branch. If I had longer then I would like to also demonstrate the signature verification features in the sample, but they are a little more complicated to set-up. If we want to continue with a sample for future release (which I would support), then I think we have potentially four approaches, probably from least to most effort:
1. Add the extra config files in the `rootfs.sh` as I did, but use an environment variable to select if it is on/off (default to off) and make this an option in the `make cc-payload`
2. Use some technology/script to modify the image after it is built during the `make cc-payload` process. We could either use something like[`cp_to_guest_img`](https://github.com/kata-containers/tests/blob/7a649220e5d9bef6260cb801ad4df74e46203b2c/lib/common.bash#L274) script, or something like [virt-customize](https://libguestfs.org/virt-customize.1.html) to edit it here before we make the container image. Again key'd by some sort of env var.
3. Use a similar approach to the previous, but doing it after the container image as been created, we could potentially even do it on a postInstall step of the operator based on an environment variable? This would avoid the need for a separate sample payload entirely, but might want us to create a new test in the operator CI for this scenario
4. Provide a generic script/template of how to custom the image that end-users can use to do their own changes and use the sample encryption & signature verification customisation as a use case and sample of how to do it. This means we wouldn't have a sample payload and 'enabling' the sample would be user directed. I'm not sure of the user experience of this as I guess it would have to be run after the 'normal' payloadImage is installed and then they might have to re-deploy the cc runtime to pick it up? I needs more thinking about by someone more clued up that me!

I think I've probably gone beyond the original scope of the issues, but thought it was worth starting the discussion.
/cc @fidencio @wainersm @bpradipt for critique

**Comment by fitzthum** [comment]
I think I would prefer something like 3 or 4. 

I think having a separate sample payload makes things a bit more complicated than they need to be. It's also a bit of a pain for a user who might have to re-deploy the operator to try out an encrypted sample image and the re-deploy it again to switch back.

Fundamentally all we are doing is switching to the `offline_fs_kbc`, updating the agent config file, and putting a file in the rootfs. For the first thing, I think we can just start building the attestation-agent with the `offline_fs_kbc` enabled for every platform (in addition to the hw-specific KBC). Then it really just becomes a case of copying a few files into the initrd.

**Comment by fitzthum** [comment]
On a related note, it looks like the non-tee config doesn't actually use the agent config file at all. I think this is sub-optimal.  IMO we should have things like Exec disabled by default in all of the configs.

**Comment by stevenhorsman** [comment]
> On a related note, it looks like the non-tee config doesn't actually use the agent config file at all. I think this is sub-optimal. IMO we should have things like Exec disabled by default in all of the configs.

I completely agree - after I raised the issue of the agent-config.toml overriding `kernel_params`, Samuel suggested 0.1 go without the agent config (and therefore exec being enabled), but we need to change this in future and tighten up our default config to be more secure.

It was on my 0.2 wishlist, but I didn't create the issue yet as I felt like https://github.com/kata-containers/kata-containers/issues/5173 (and maybe https://github.com/kata-containers/kata-containers/issues/5178) were blockers we needed to get done first and might influence the approach we take with this default.

Thanks for the feedback Tobin!

**Comment by Xynnn007** [comment]
Sry and I have a question out of scope. Do we have any plan to publish a community-version image that do not use KBSes based on filesystem? like the coming cc-kbc.

**Comment by stevenhorsman** [comment]
> Sry and I have a question out of scope. Do we have any plan to publish a community-version image that do not use KBSes based on filesystem? like the coming cc-kbc.

That's a good question. So the creation of the offline_fs_kbc approach was intended to be a short term thing so that a user who was just investigating confidential-containers could try out support for encrypted images/image signatures without having to fully understand and stand up to full infrastructure like a remote KBS. As mentioned above, we intend to replace this with a script/easier documentation to let users prepare their own, so we'd look to remove even the image with the pre-configured offline_fs_kbc in.

With respect to whether we'd publish a community image with something like cc-kbc, my thoughts are:
- Could this just be the 'normal' image now we can pass through the `aa_kbc_params` through the kata-containers config?, or is there anything else unique that would need to be included in the image?
- I think if we needed a custom image for a new remote-based KBS/KBC then we'd need to have some good doc of how to set this up and it's probably a different use case than the offline_fs_kbc image which doesn't need confidential hardware and would be closer to actual confidential computing?

Let me know your thoughts and we can always raise this further to the wider community. Thanks!

**Comment by Xynnn007** [comment]
Thanks for your detailed explanation @stevenhorsman .

> * Could this just be the 'normal' image now we can pass through the `aa_kbc_params` through the kata-containers config?, or is there anything else unique that would need to be included in the image?

IMO, it should be a 'normal' image with a newest stable `kata-agent`, `attestation-agent` (with feature `cc-kbc` or some specific kbc). The reasons for this are:
* A user can directly deploy kata-cc environment
* Helps to provide a fixed measurement for `rootfs`. This is also the core idea I want to express. Currently when we do Remote Attestation between `AA` and `AS`, the evidence of rootfs is included in kernel-parameter (using the mechanism of dm-verity to verify). However, the digest of image can vary from one build to another (for not support reproducible build), so it is hard for a user to have a fixed 'reference value' of rootfs. A way to deal with this is CoCo community to publish an 'official/community' version, together with its expected digest (s.t. reference value of the rootfs), thus helping AS to do the remote attestation when users use this out-of-the-box image.


> * I think if we needed a custom image for a new remote-based KBS/KBC then we'd need to have some good doc of how to set this up and it's probably a different use case than the offline_fs_kbc image which doesn't need confidential hardware and would be closer to actual confidential computing?

Yes, it is.

**Comment by stevenhorsman** [comment]
> Thanks for your detailed explanation @stevenhorsman .
> 
> > * Could this just be the 'normal' image now we can pass through the `aa_kbc_params` through the kata-containers config?, or is there anything else unique that would need to be included in the image?
> 
> IMO, it should be a 'normal' image with a newest stable `kata-agent`, `attestation-agent` (with feature `cc-kbc` or some specific kbc). The reasons for this are:
> 
> * A user can directly deploy kata-cc environment
> * Helps to provide a fixed measurement for `rootfs`. This is also the core idea I want to express. Currently when we do Remote Attestation between `AA` and `AS`, the evidence of rootfs is included in kernel-parameter (using the mechanism of dm-verity to verify). However, the digest of image can vary from one build to another (for not support reproducible build), so it is hard for a user to have a fixed 'reference value' of rootfs. A way to deal with this is CoCo community to publish an 'official/community' version, together with its expected digest (s.t. reference value of the rootfs), thus helping AS to do the remote attestation when users use this out-of-the-box image.

Great, thanks for the confirmation and additional info. I have another follow-up based on this point:
>  IMO, it should be a 'normal' image with a newest stable `kata-agent`, `attestation-agent` (with feature `cc-kbc` or some specific kbc)

At the moment we just include a single kbc feature when we build the image and the `AA_KBC` param value is used to set-up some config as well. Do you think it's viable to install all the KBC features in the attestation-agent that we include in the rootfs image (and I guess any required config files e.g. for the offline KBCs) and then just use the `aa_kbc_params` parameter to pick which one is used at 'runtime'. If so I guess we could try moving to this for the default image and then we'd only need the single measurement for the rootfs?
Alternatively, if this isn't possible, should we be considering building the 'community image' for each KBS/KBC (including the default kata-containers config to help with the measurement of that?)?

There is probably things I've missed here, so please let me know! Thanks.

/cc @fidencio - as you are a key party here.

**Comment by Xynnn007** [comment]
> Do you think it's viable to install all the KBC features in the attestation-agent that we include in the rootfs image (and I guess any required config files e.g. for the offline KBCs) and then just use the `aa_kbc_params` parameter to pick which one is used at 'runtime'. If so I guess we could try moving to this for the default image and then we'd only need the single measurement for the rootfs?

I agree with the one-measurement way. As customized could be impled in future if needed.

**Comment by fitzthum** [comment]
Not exactly sure about the future of the sample image, but I agree that the "normal" image should eventually ship with the `cc-kbc`. We could also build some of the other KBCs for compatibility although note that these KBCs require other changes to the initrd. For instance, sev kbcs require the efi-secret kernel module and the offline_fs_kbc requires the files to be in the filesystem. So we will still have a few puzzles to solve if we want a totally generic image.
